cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/DefaultEndpointPluginContext.java

```diff
@@ -0,0 +1,123 @@
+import co.cask.cdap.api.plugin.PluginClass;
+import co.cask.cdap.api.plugin.PluginProperties;
+import co.cask.cdap.api.plugin.PluginPropertyField;
+import co.cask.cdap.api.plugin.PluginSelector;
+import co.cask.cdap.common.ArtifactNotFoundException;
+import co.cask.cdap.internal.app.runtime.artifact.ArtifactDescriptor;
+import co.cask.cdap.internal.app.runtime.artifact.ArtifactRepository;
+import co.cask.cdap.internal.app.runtime.plugin.PluginInstantiator;
+import co.cask.cdap.internal.app.runtime.plugin.PluginNotExistsException;
+import co.cask.cdap.proto.Id;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * An implementation of {@link EndpointPluginContext} that uses {@link PluginInstantiator}
+ */
+public class DefaultEndpointPluginContext implements EndpointPluginContext {
+
+  private final ArtifactRepository artifactRepository;
+  // this is the namespace of the artifact
+  private final Id.Namespace deployNamespace;
+
+  private final PluginInstantiator pluginInstantiator;
+  private final Id.Artifact artifactId;
+
+  public DefaultEndpointPluginContext(Id.Namespace deployNamespace, ArtifactRepository artifactRepository,
+                                      PluginInstantiator pluginInstantiator, Id.Artifact artifactId) {
+    this.deployNamespace = deployNamespace;
+    this.artifactRepository = artifactRepository;
+    this.pluginInstantiator = pluginInstantiator;
+    this.artifactId = artifactId;
+  }
+
+  @Override
+  public <T> Class<T> newPluginInstance(String pluginType, String pluginName) {
+   return newPluginInstance(pluginType, pluginName, PluginProperties.builder().build());
+  }
+
+  @Override
+  public <T> Class<T> newPluginInstance(String pluginType, String pluginName,
+                                        PluginProperties pluginProperties) {
+    Plugin plugin;
+    try {
+      plugin = findPlugin(artifactId, pluginType, pluginName, pluginProperties, new PluginSelector());
+    } catch (PluginNotExistsException e) {
+      // Plugin not found, hence return null
+      return null;
+    } catch (ArtifactNotFoundException e) {
+      // this shouldn't happen, it means the artifact for this app does not exist.
+      throw new IllegalStateException(
+        String.format("Application artifact '%s' no longer exists. Please check if it was deleted.",
+                      artifactId));
+    }
+
+    try {
+      Class<T> cls = pluginInstantiator.loadClass(plugin);
+      return cls;
+    } catch (IOException e) {
+      // If the plugin jar is deleted without notifying the artifact service.
+      return null;
+    } catch (ClassNotFoundException e) {
+      // Shouldn't happen
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * clone of {@link co.cask.cdap.internal.app.DefaultPluginConfigurer#findPlugin(String, String,
+   * String, PluginProperties, PluginSelector)}
+   */
```

Comment:
this code is a duplicate. one way to avoid duplicate is to move findPlugin implementation to `AbstractPluginHelper`, then make `DefaultEndpointPluginContext` extend that class.

`DefaultPluginConfigurer` currently extends `DefaultDatasetConfigurer` we can rather have it implement dataset configurer and delegate to `DefaultDatasetConfigurer` and then extend `AbstractPluginHelper` , let me know if we should do that instead.